SW-855: Wiki rework

mohjohfox opened the pull request.

mohjohfox committed:
SW-855 Convert lib from java to kotlin

mohjohfox committed:
SW-855 Added to-do comments

mohjohfox committed:
SW-855 Added functions  to persist topics and update them

mohjohfox committed:
SW-855 Added port functions for creating and updating entries

mohjohfox committed:
SW-855 Use correct update function for prepared statements

mohjohfox committed:
SW-855 Added template methods for query entities

Added empty functions for:
- getting wikiType
- getting topic
- getting entity

mohjohfox committed:
SW-855 Get topic entity from db via id

mohjohfox committed:
SW-855 Get entry entity from db via id

mohjohfox committed:
SW-855 Added getParent from child_id query and fixed missing variable in model

- Added child_ids to TopicEntity model

mohjohfox committed:
SW-855 Update child_ids on topic update

mohjohfox committed:
SW-855 Added create and update function to service class

mohjohfox committed:
SW-855 Added functions to create and update entry and changed child_id list to mutable list in topic entity

SteffenBurgert committed:
SW-855: small refactoring

mohjohfox committed:
SW-855 feat(api)!: Change function api

mohjohfox committed:
SW-855 feat: Return generated key when adding new wiki entry

mohjohfox committed:
SW-855 build: Set snapshot version

mohjohfox committed:
SW-855 ci: Add actions for feature branches

mohjohfox committed:
SW-855 ci: Permit gradle

mohjohfox committed:
SW-855 ci: Fix gradle error

mohjohfox committed:
Merge branch 'main' into SW-855-database-statements

mohjohfox committed:
SW-855 build: Fix build

mohjohfox committed:
SW-855 build: Fix build and added first test cases

mohjohfox committed:
SW-855 fix: Fix serveral sql statements and introduced RowMapper for query function

mohjohfox committed:
SW-855 testing: Add test cases for sql queries

mohjohfox committed:
SW-855 build: Set version to 1.0.0-SNAPSHOT

Mirkonator committed:
SW-855 template for wiki insert witch id returning as example on createTopic.

mohjohfox committed:
SW-855 Introduce returning id in sql queries

mohjohfox committed:
SW-855 fix: Use prepared statements

mohjohfox committed:
SW-855 testing: Fix typo in package name

mohjohfox committed:
SW-855 feat: Check for correct types when creating entries and link topic to parent topic

mohjohfox committed:
SW-855 testing: Add test cases and adapt test data

mohjohfox committed:
SW-855 feat: Add delete function for topics and entries

mohjohfox committed:
SW-855 testing: Add delete test cases

mohjohfox committed:
SW-855 feat: Introduce WikiException and throw it if some check is failing

mohjohfox committed:
SW-855 testing: Add test cases for throwing exceptions

mohjohfox committed:
SW-855 feat: Add poll support and introduce hasAccess function to check access

SteffenBurgert commented on build.gradle.kts:
```diff
@@ -1,78 +1,78 @@
-	withSourcesJar()
-	//withJavadocJar()
+    withSourcesJar()
+    //withJavadocJar()
```
Kann das weg?

mohjohfox replied:
Zukünftig eventulle im Einsatz. Würde es mal drin lassen

SteffenBurgert commented on build.gradle.kts:
```diff
@@ -1,78 +1,78 @@
 repositories {
-	mavenCentral()
+    mavenCentral()
 }
```
Ist schon in Zeile 24

SteffenBurgert commented on build.gradle.kts:
```diff
@@ -1,78 +1,78 @@
+    implementation("org.springframework.boot:spring-boot-starter-jdbc")
+    implementation("org.springframework.boot:spring-boot-autoconfigure:3.3.2")
+    implementation("javax.annotation:javax.annotation-api:1.2-b01")
+    implementation("com.google.code.gson:gson:2.8.9")
```
Würde evtl. schauen, nur Jackson zu verwenden oder was meinst du? Ich glaube, das verwendet man ja für JSON

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/WikiAdapter.kt:
```diff
@@ -0,0 +1,149 @@
+                    val id = wikiSqlAdapter.createTopic(TopicEntity(null, topic, mutableListOf()))
+                    val childIds = parent.childIds.toMutableList()
+                    childIds.add(id)
+                    wikiSqlAdapter.updateTopic(parent.copy(childIds = childIds))
```
Kannst überlegen, ob du das schreiben willst:
wikiSqlAdapter.updateTopic(
    parent.copy(childIds = parent.childIds.toMutableList().apply { add(id) })
)

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/WikiAdapter.kt:
```diff
@@ -0,0 +1,149 @@
+                        topic.childIds.forEach { wikiSqlAdapter.deleteEntry(it) }
+                        val childIds = parent.childIds.toMutableList()
+                        childIds.remove(id)
+                        wikiSqlAdapter.updateTopic(parent.copy(childIds = childIds))
```
Alternativ:
wikiSqlAdapter.updateTopic(parent.copy(
      childIds = parent.childIds.toMutableList().apply { remove(id) })
)

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/WikiAdapter.kt:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    @Transactional
+    override fun createTopic(parentId: Int, topic: String, hasAccess: () -> Boolean): Int {
```
Was findest du besser? Mein Vorschlag wäre:

        if (hasAccess.invoke().not()) {
            throw WikiAccessDeniedException("Access denied for creating topic for parentId $parentId")
        }

        val wikiType = wikiSqlAdapter.getWikiType(parentId)
        val parent = wikiSqlAdapter.getTopic(parentId)

        if (wikiType != WikiType.TOPIC || parent == null) {
            throw WikiNotFoundException("No topic found to create a child topic with parentId $parentId")
        }

        val id = wikiSqlAdapter.createTopic(TopicEntity(null, topic, listOf()))
        wikiSqlAdapter.updateTopic(
            parent.copy(childIds = parent.childIds.toMutableList().apply { add(id) })
        )
        return id

Dadurch haben wir keine so große Verschachtelung. Bin mir da noch unsicher. Glaube aber das wäre besser, da du 3 ifs ineinander schachtelst.

Könnte man dann auch für die weiteren Methoden anpassen.

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/WikiAdapter.kt:
```diff
@@ -0,0 +1,149 @@
+                    val entryId = wikiSqlAdapter.createEntry(EntryEntity(null, headline, body))
+                    val childIds = parent.childIds.toMutableList()
+                    childIds.add(entryId)
+                    wikiSqlAdapter.updateTopic(parent.copy(childIds = childIds))
```
evtl mit apply?

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/WikiAdapter.kt:
```diff
@@ -0,0 +1,149 @@
+                if (parent != null) {
+                    val childIds = parent.childIds.toMutableList()
+                    childIds.remove(id)
+                    wikiSqlAdapter.updateTopic(parent.copy(childIds = childIds))
```
evtl mit apply?

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/WikiPollAdapter.kt:
```diff
@@ -0,0 +1,65 @@
+
+    @Transactional
+    override fun savePoll(topicId: Int, poll: PollEntity, hasAccess: () -> Boolean): Int {
+        if (wikiSqlAdapter.getWikiType(topicId) != null) {
```
Was hältst du davon:
        val topic = wikiSqlAdapter.getTopic(topicId)

        check(wikiSqlAdapter.getWikiType(topicId) != null && topic != null) {
            throw IllegalArgumentException("No topic found with id $topicId")
        }

        val id = pollSqlAdapter.savePoll(poll)
        wikiSqlAdapter.updateTopic(topic.copy(childIds = topic.childIds + id))
        return id

Ich glaube das es besser ist, wenn wir nicht so verschachtelt sind.

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/dto/EntryEntity.kt:
```diff
@@ -0,0 +1,7 @@
+package de.simpletactics.wiki.lib.adapter.dto
+
+data class EntryEntity(
+    val id: Int? = null,
```
Das = null kann weg, dadurch dass wir danach keinen Default haben kann das nie verwendet werden. Außer man würde bei dem EntryEntity Aufruf folgendes schreiben:
EntryEntity(headline = "headline", htmlEntry = "entry")
davon würde ich aber abraten. Man könnte die ID ans Ende machen, das fände ich aber komisch. Für mich ist das immer das Erste.

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/dto/TopicEntity.kt:
```diff
@@ -0,0 +1,7 @@
+package de.simpletactics.wiki.lib.adapter.dto
+
+data class TopicEntity(
+    val id: Int? = null,
```
= null kann man sich sparen. (siehe EntryEntity)

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/dto/poll/Date.kt:
```diff
@@ -0,0 +1,38 @@
+import java.time.ZonedDateTime
+import java.time.format.DateTimeFormatter
+
+object Date {
```
Mann könnte ich überlegen Extension Funktions zu schreiben. Wäre eine Stilfrage.

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/dto/poll/PollEntity.kt:
```diff
@@ -0,0 +1,22 @@
+data class PollEntity(
+    val id: Int?,
+    val question: String,
+    val description: String = "",
```
Default ist unnötig, außer es werden nur id, question und pollentries spezifisch als Übergabeparameter. Mann könnte auch pollEntries mit emptyList initialisieren, aber ich glaube, das macht nicht viel Sinn. Evtl. nullable?

SteffenBurgert replied:
Vielleicht description und pollEntries tauschen

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/dto/poll/PollEntity.kt:
```diff
@@ -0,0 +1,22 @@
+data class PollModel(
+    val id: Int?,
+    val question: String,
+    val description: String = "",
```
Hier ist der Default auch unnötig. (Siehe PollEntity) vielleicht nullable bei einem leeren String?

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/dto/poll/PollEntryEntity.kt:
```diff
@@ -0,0 +1,12 @@
+import de.simpletactics.model.poll.Vote
+
+data class PollEntryEntity(
+    @JsonProperty("pollOption")
+    val pollOption: PollOption,
+    @JsonProperty("votes")
+    val votes: MutableList<Vote>,
```
Braucht man die Annotation, wenn die gleich geschrieben sind. Das wird mit Jackson gemappt und ich meine Jackson braucht das nicht. Ich kann mich aber auch Irren.
Bzw. Wenn wir uns Dtos erzeugen können wir das auch ignoreren.

SteffenBurgert replied:
Wird scheibar für das Mapping mit der db benötigt

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/dto/poll/PollOption.kt:
```diff
@@ -0,0 +1,10 @@
+import com.fasterxml.jackson.annotation.JsonProperty
+
+data class PollOption(
+    @JsonProperty("uuid")
+    val uuid: String,
+    @JsonProperty("text")
+    val text: String,
```
siehe PollEntryEntity

SteffenBurgert replied:
Wird scheibar für das Mapping mit der db benötigt

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/dto/poll/Vote.kt:
```diff
@@ -0,0 +1,12 @@
+import com.fasterxml.jackson.annotation.JsonProperty
+
+data class Vote(
+    @JsonProperty("userId")
+    val userId: Int,
+    @JsonProperty("date")
+    val date: String,
+    @JsonProperty("option")
+    val option: PollVoteEnum,
```
siehe PollEntryEntity

SteffenBurgert replied:
Wird scheibar für das Mapping mit der db benötigt

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/persistence/PollSqlAdapter.kt:
```diff
@@ -0,0 +1,179 @@
+            pollEntry.first()
+        } else {
+            null
+        }
```
return jdbc.query("select * from wiki_poll where id = ?", PollMapper(), id).first() ?: null

Sollte das Gleiche machen.

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/persistence/PollSqlAdapter.kt:
```diff
@@ -0,0 +1,179 @@
+    fun savePoll(
+        poll: PollEntity,
+    ): Int {
+        val date = if (poll.date != null) "'${poll.date}'" else null
```
Kann man denke ich Inlinen.

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/persistence/PollSqlAdapter.kt:
```diff
@@ -0,0 +1,179 @@
+            getPoll(poll.id)
+        } else {
+            throw WikiNotFoundException("Couldn't update poll. Poll id is null")
+        }
```
Würde versuchen auf else zu verzichten, wenn es geht: 
       if (poll.id != null) {
            val pollFromDb = getPoll(poll.id)

            if (pollFromDb != null) {
                updatingEntries(poll, pollFromDb)
            }
            return getPoll(poll.id)
        }

        throw WikiNotFoundException("Couldn't update poll. Poll id is null")

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/persistence/PollSqlAdapter.kt:
```diff
@@ -0,0 +1,179 @@
+        pollModel: PollEntity,
+    ) {
+        val oldPoll = getPoll(pollModel.id!!)
+
+        if (oldPoll != null) {
+            pollModel.pollEntries.forEach {
+                val option = getUserVoteOption(it)
+                val date = Date.getDateAsString()
+
+                jdbc.execute(
+                    "UPDATE wiki_poll set data = (" +
+                            "SELECT JSONB_AGG(" +
+                            "CASE when d.value @? '$.pollOption.uuid ? (@ == \"${it.pollOption.uuid}\")' " +
+                            "THEN " +
+                            "CASE when d.value @? '$.votes ?(@.userId == $userId)' " +
+                            "THEN jsonb_set(d.value, '{votes}', " +
+                            "(SELECT JSONB_AGG( " +
+                            "CASE when dd.value @? '$ ?(@.userId == $userId)' AND NOT dd.value -> 'option' @> '\"$option\"'::jsonb " +
+                            "THEN jsonb_set(jsonb_set(dd.value," +
+                            "'{option}', '\"$option\"')" +
+                            ", '{date}', '\"$date\"') " +
+                            "ELSE dd.value " +
+                            "END) " +
+                            "FROM jsonb_array_elements(jsonb_path_query_array(d.value, '$.votes[*]')) dd)" +
+                            " ) " +
+                            "ELSE jsonb_insert(d.value, '{votes, -1}', '{" +
+                            "\"userId\": $userId," +
+                            "\"date\": \"$date\"," +
+                            "\"option\": \"$option\"" +
+                            "}', true) " +
+                            "END " +
+                            "ELSE d.value " +
+                            "END) " +
+                            "FROM jsonb_array_elements(data) d) " +
+                            "where id = ${pollModel.id}",
+                )
+            }
+        } else {
+            log.warn("Could not set poll votes because there is no poll with id ${pollModel.id}")
+        }
```
Würde hierzu tendieren:
        if (pollModel.id != null && getPoll(pollModel.id) == null) {
            log.warn("Could not set poll votes because there is no poll with id ${pollModel.id}")
            return
        }

        pollModel.pollEntries.forEach {
            val option = getUserVoteOption(it)
            val date = Date.getDateAsString()

            jdbc.execute(
                """UPDATE wiki_poll set data = (
                        SELECT JSONB_AGG(
                        CASE when d.value @? '$.pollOption.uuid ? (@ == \"${it.pollOption.uuid}\")'
                        THEN
                        CASE when d.value @? '$.votes ?(@.userId == $userId)'
                        THEN jsonb_set(d.value, '{votes}',
                        (SELECT JSONB_AGG(
                        CASE when dd.value @? '$ ?(@.userId == $userId)' AND NOT dd.value -> 'option' @> '\"$option\"'::jsonb
                        THEN jsonb_set(jsonb_set(dd.value,
                        '{option}', '\"$option\"')
                        , '{date}', '\"$date\"')
                        ELSE dd.value
                        END)
                        FROM jsonb_array_elements(jsonb_path_query_array(d.value, '$.votes[*]')) dd)
                         )
                        ELSE jsonb_insert(d.value, '{votes, -1}', '{
                        \"userId\": $userId,
                        \"date\": \"$date\",
                        \"option\": \"$option\"
                        }', true)
                        END
                        ELSE d.value
                        END)
                        FROM jsonb_array_elements(data) d)
                        where id = ${pollModel.id}""",
            )
        }

Wenn wir drei Anführungszeichen schreiben, dann haben noch Unterstützung von der IDE.

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/persistence/WikiSqlAdapter.kt:
```diff
@@ -0,0 +1,102 @@
+    private val jdbc: JdbcTemplate,
+) {
+
+    fun getWikiType(id: Int): WikiType? {
+        val result = jdbc.queryForList("SELECT type FROM wiki WHERE id = ?", id)
+        return if (result.size == 1 && result.first().containsKey("type")) {
+            val typeAsString = result.first()["type"].toString()
+            WikiType.valueOf(typeAsString)
+        } else {
+            null
+        }
+    }
```
Weiß nicht was besser ist:
fun getWikiType(id: Int): WikiType? {
    val result = jdbc.queryForList("SELECT type FROM wiki WHERE id = ?", id)

    check(result.size == 1 && result.first().containsKey("type")) {
        "Invalid result for id $id: Expected exactly one row with a 'type' column"
    }

    return WikiType.valueOf(result.first()["type"].toString())
}

Hier würden wir noch einen Fehler werfen.

SteffenBurgert replied:
Können wir auch mehr als 1 zurückbekommen? Fall nicht könnte man das glaube ich auch anders schreiben. Ich versuche mittlerweile else so gut es geht zu vermeiden.

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/persistence/mapper/EntityMapper.kt:
```diff
@@ -0,0 +1,38 @@
+import de.simpletactics.wiki.lib.adapter.dto.poll.PollModel
+
+
+fun PollEntity.toModel(): PollModel {
```
toPollModel. Wäre besser, damit man das gleich lesen kann.

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/persistence/mapper/EntityMapper.kt:
```diff
@@ -0,0 +1,38 @@
+    )
+}
+
+private fun getOptionSelectionFromVotes(
+    votes: List<Vote>,
+    userId: Int,
+): PollVoteEnum {
+    return votes.filter { it.userId == userId }.map { it.option }.firstOrNull()
+        ?: PollVoteEnum.FALSE
+}
```
Wird nicht verwendet uns ist private. Kann das weg?

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/persistence/mapper/RowMapper.kt:
```diff
@@ -0,0 +1,67 @@
+
+class TopicRowMapper : RowMapper<TopicEntity> {
+
+    override fun mapRow(rs: ResultSet, rowNum: Int): TopicEntity? {
```
? kann weg
würde rs zu results ändern

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/persistence/mapper/RowMapper.kt:
```diff
@@ -0,0 +1,67 @@
+
+class EntryMapper : RowMapper<EntryEntity> {
+
+    override fun mapRow(rs: ResultSet, rowNum: Int): EntryEntity? {
```
? kann weg
würde rs zu results ändern

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/persistence/mapper/RowMapper.kt:
```diff
@@ -0,0 +1,67 @@
+}
+
+class IdMapper : RowMapper<Int> {
+    override fun mapRow(rs: ResultSet, rowNum: Int): Int? {
```
? kann weg
würde rs zu results ändern

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/persistence/mapper/RowMapper.kt:
```diff
@@ -0,0 +1,67 @@
+}
+
+class PollMapper : RowMapper<PollEntity> {
+    override fun mapRow(rs: ResultSet, rowNum: Int): PollEntity? {
```
? kann weg
würde rs zu results ändern

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/persistence/mapper/RowMapper.kt:
```diff
@@ -0,0 +1,67 @@
+}
+
+class PollOpenBooleanMapper : RowMapper<Boolean> {
+    override fun mapRow(rs: ResultSet, rowNum: Int): Boolean? {
```
? kann weg
würde rs zu results ändern

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/persistence/mapper/RowMapper.kt:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    private fun parseJsonToPollEntity(json: String): List<PollEntryEntity> {
+        val mapper = ObjectMapper()
```
Tendiere zu einem globalen Object Mapper

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/model/Right.kt:
```diff
@@ -0,0 +1,10 @@
+    val edit: Boolean,
+    val create: Boolean,
+    val write: Boolean,
+    val admin: Boolean,
+    val mod: Boolean,
```
Raus mit den Rollen. Evtl neue Klasse mit Rollen. Glaube aber dass wir die auch weglassen können. Man kann die Später denke auch mal asu der db Laden.

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/model/SubjectEntry.kt:
```diff
@@ -0,0 +1,11 @@
+package de.simpletactics.wiki.lib.model
+
+data class SubjectEntry(
+    val wikiType: WikiType = WikiType.TOPIC,
```
Default ist unnötig, man muss immer alles angeben.

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/services/JsonParser.kt:
```diff
@@ -0,0 +1,26 @@
+import org.springframework.stereotype.Component
+
+@Component
+class JsonParser(
+    private var gson: Gson,
+) {
+
+    init {
+        val gsonBuilder = GsonBuilder()
+        gsonBuilder.setPrettyPrinting()
+        gson = gsonBuilder.create()
+    }
+
+    fun <T> parsetoJson(objects: T): String {
+        return gson.toJson(objects)
+    }
+
+    fun <T> parseToClass(json: String, classType: T): T {
+        return gson.fromJson(json, classType!!::class.java)
+    }
+
+}
```
Würde statt globalem GSON eher einen Globalen ObjectMapper mit Jackson für JSON verwenden. Dann könnte man auch von GSON weg, sodass wir nur eine Mapping Lib verwenden.

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/services/port/WikiRightsPort.kt:
```diff
@@ -0,0 +1,25 @@
+
+interface WikiRightsPort {
+
+    fun getWikiBerechtigung(username: String, wikiId: Int): Right?
```
Glaube, wir haben gesagt das ? kann weg

SteffenBurgert commented on src/test/kotlin/de/simpletactics/wiki/lib/adapter/WikiAdapterTest.kt:
```diff
@@ -0,0 +1,150 @@
+@SpringBootTest
+@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
+@ActiveProfiles("testing")
+class WikiAdapterTest {
```
Tendiere zu FunSpec

SteffenBurgert reviewed (commented):
Tests habe ich noch nicht durch.

SteffenBurgert committed:
SW-855: implemented / refactored integration tests (WikiAdapterTest)

mohjohfox committed:
SW-855 testing: Annotate FunSpecIT class with spring annotations

mohjohfox committed:
SW-855 testing: Add wiki_poll.sql

mohjohfox committed:
SW-855 testing: Adapt test classes

- Adapt poll test cases for new sql scripts
- Move test classes in correct packages
-

mohjohfox committed:
SW-855 testing: Call init script only once if container is started

mohjohfox committed:
SW-855 testing: Updated kotest version

SteffenBurgert committed:
SW-855: refactored WikiAdapter

mohjohfox committed:
feat: Verify query results

SteffenBurgert committed:
SW-855: refactored Adapter, Mapper and created new ObjectMapper

SteffenBurgert committed:
Merge remote-tracking branch 'origin/SW-855-database-statements' into SW-855-database-statements

SteffenBurgert committed:
SW-855: changed PollModel order

mohjohfox committed:
SW-855 refactoring: Remove unnecessary model

mohjohfox committed:
fix: Add tests and fixed poll date insert

mohjohfox committed:
SW-855 feat: Add topic for child to WikiPort and move verify function into extension function class

mohjohfox committed:
SW-855 testing: Add WikiAccessDeniedException test

mohjohfox committed:
SW-855 refactoring: Remove null initialization and removed unused method

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/persistence/PollSqlAdapter.kt:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    fun updatePoll(poll: PollEntity): PollEntity? {
+        if (poll.id == null) throw WikiNotFoundException("Couldn't update poll. Poll id is null")
```
Mit verify abändern? Könnte dann aber sein, dass man verify mit einem booleschen Ausdruck abändern müsste.
Muss aber nicht sein.

mohjohfox replied:
Wir werfen jetzt eine `IllegalStateException` wenn die Id null ist. Das sollte im Normfall nicht auftreten. Würde die `WikiNotFoundException` werfen, wenn technisch der Zugriff korrekt ist und es den Eintrag mit der id nicht gibt. Eine null id sollte einen anderen Fehler wie `IllegalStateException` werfen oder was meinst du?

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/persistence/PollSqlAdapter.kt:
```diff
@@ -0,0 +1,172 @@
+    ) {
+        if (pollModel.id != null && getPoll(pollModel.id) == null) {
+            // TODO: throw Exception???
+            log.warn("Can not set poll votes because there is no poll with id ${pollModel.id}")
```
Offenes TODO?

mohjohfox replied:
Done

SteffenBurgert commented on src/main/kotlin/de/simpletactics/wiki/lib/adapter/persistence/PollSqlAdapter.kt:
```diff
@@ -0,0 +1,172 @@
+    private fun getUserVoteOption(pollEntryEntity: PollEntryEntity) =
+        pollEntryEntity.votes[0].option
+
+    /*
+     * jsonb_array_elements() teilt das Array im jsonb auf und gibt die Einzelteile aus.
+     * jsonb_agg() nutzt fügt das aufgeteilte Array wieder zusammen
+     * Über CASE wird entschieden, ob die Einzelteile des Array so bleiben oder
+     * vor dem Zusammensetzen mit jsonb_set() geändert werden.
+     */
```
Wollen wir den auf Englisch schreiben?

SteffenBurgert commented on src/test/kotlin/de/simpletactics/wiki/lib/adapter/persistence/poll/PollPropertyChangeAdapterTest.kt:
```diff
@@ -0,0 +1,32 @@
+            pollPort.isPollOpen(6) { true } shouldBe false
+        }
+
+        test("close all open polls") {
```
close all open expired polls

mohjohfox replied:
Done

SteffenBurgert commented on src/test/kotlin/de/simpletactics/wiki/lib/adapter/persistence/poll/SavePollAdapterTest.kt:
```diff
@@ -0,0 +1,116 @@
+        null,
+    )
+
+    // Nur im Test hinterlegen
```
kann weg

mohjohfox replied:
Done

SteffenBurgert commented on src/test/kotlin/de/simpletactics/wiki/lib/adapter/persistence/poll/SavePollAdapterTest.kt:
```diff
@@ -0,0 +1,116 @@
+    val mockedNotPresentPollId = 999
+    val mockedPresentDatabasePollId = 5
+
+    // Nur im Test hinterlegen
```
kann weg

mohjohfox replied:
Done

SteffenBurgert commented on src/test/resources/application-testing.properties:
```diff
@@ -0,0 +1,4 @@
+spring.datasource.driver-class-name=org.testcontainers.jdbc.ContainerDatabaseDriver
+#spring.datasource.url=jdbc:tc:postgresql:15.3-alpine:///integration-tests-db
```
kann das weg?

mohjohfox replied:
Done

mohjohfox committed:
SW-855 fix: Add checks before modifing polls

- check if id is not null before updating
- verify that id belongs to a poll before setting votes

mohjohfox committed:
SW-855 testing: Remove comments from test cases and name test correctly

mohjohfox committed:
build: Use spring-boot-starter-data-jdbc

mohjohfox committed:
feat: Rollback on SqlException

- Introduce SitaTransactional annotation
- Wrap spring transaction with custom rollback logic

SteffenBurgert committed:
SW-855: fixed end poll

mohjohfox committed:
fix: Remove poll id from parent when deleting poll

SteffenBurgert committed:
SW-855: fixed test delete poll

mohjohfox commented on build.gradle.kts:
```diff
@@ -1,78 +1,74 @@
 group = "de.simpletactics"
-version = "0.0.2"
+version = "1.0.0-SNAPSHOT"
```
Remove snapshot version before merging

mohjohfox committed:
SW-855 build: Remove snapshot from version number
